File: packages/types/src/harness-session-id.test.ts
import { describe, it, expect } from 'vitest'
import { HarnessError } from './errors.js'
import { HARNESS_IDS } from './harness.js'
import {
  decodeSessionIdSegment,
  encodeSessionIdSegment,
  formatSessionId,
  isSessionId,
  parseSessionId,
} from './harness-session-id.js'

/** Every rejection must be a typed `invalid_session_id`, not a bare Error —
 *  the gateway maps the code to HTTP 400. */
function expectInvalid(fn: () => unknown) {
  expect(fn).toThrow(HarnessError)
  try {
    fn()
  } catch (err) {
    expect((err as HarnessError).code).toBe('invalid_session_id')
  }
}

describe('parseSessionId', () => {
  it('splits on the FIRST colon so native ids may contain colons', () => {
    expect(parseSessionId('grok-build:sess_01HZX:part:two')).toEqual({
      harnessId: 'grok-build',
      nativeSessionId: 'sess_01HZX:part:two',
    })
  })

  it('accepts every harness id in the enum', () => {
    for (const harnessId of HARNESS_IDS) {
      expect(parseSessionId(`${harnessId}:abc`).harnessId).toBe(harnessId)
    }
  })

  it('keeps path-shaped legacy Claude native ids intact', () => {
    expect(parseSessionId('claude-code:-home-rivet-project/a1b2c3d4')).toEqual({
      harnessId: 'claude-code',
      nativeSessionId: '-home-rivet-project/a1b2c3d4',
    })
  })

  it('rejects surrounding whitespace instead of silently trimming', () => {
    expectInvalid(() => parseSessionId(' claude-code:abc'))
    expectInvalid(() => parseSessionId('claude-code:abc '))
    expectInvalid(() => parseSessionId('\tclaude-code:abc\n'))
    // ...but whitespace INSIDE the opaque native id is preserved, not rejected.
    expect(parseSessionId('claude-code:a b').nativeSessionId).toBe('a b')
  })

  it('rejects unknown harness ids and legacy prefixes', () => {
    expectInvalid(() => parseSessionId('claude:abc'))
    expectInvalid(() => parseSessionId('cc:abc'))
    expectInvalid(() => parseSessionId('CLAUDE-CODE:abc'))
    expectInvalid(() => parseSessionId('task:t_123'))
  })

  it('rejects an empty native id, an empty harness id, and a bare native id', () => {
    expectInvalid(() => parseSessionId('claude-code:'))
    expectInvalid(() => parseSessionId(':abc'))
    expectInvalid(() => parseSessionId('a1b2c3d4-uuid'))
    expectInvalid(() => parseSessionId(''))
  })
})

describe('formatSessionId', () => {
  it('round-trips with parseSessionId, colons and slashes included', () => {
    const cases: Array<[(typeof HARNESS_IDS)[number], string]> = [
      ['claude-code', 'a1b2c3d4-5e6f-7890-abcd-ef1234567890'],
      ['claude-code', '-home-rivet-rivetOS/a1b2c3d4-5e6f'],
      ['grok-build', 'sess_01HZX:2:3'],
      ['kimi-code', 'c7f2-uuid'],
      ['hermes', '9b41-uuid'],
    ]
    for (const [harnessId, nativeSessionId] of cases) {
      const id = formatSessionId(harnessId, nativeSessionId)
      expect(id).toBe(`${harnessId}:${nativeSessionId}`)
      expect(parseSessionId(id)).toEqual({ harnessId, nativeSessionId })
    }
  })

  it('rejects an empty or trailing-whitespace native id', () => {
    expectInvalid(() => formatSessionId('claude-code', ''))
    expectInvalid(() => formatSessionId('claude-code', 'abc '))
    // Asymmetric by contract: the trim check is on the whole SessionId, and a
    // LEADING space sits after the separator, so `claude-code: abc` is a legal
    // (if ugly) canonical id — the native half is opaque.
    expect(formatSessionId('claude-code', ' abc')).toBe('claude-code: abc')
  })
})

describe('isSessionId', () => {
  it('narrows valid ids and rejects invalid ones without throwing', () => {
    expect(isSessionId('hermes:9b41-uuid')).toBe(true)
    expect(isSessionId('claude:9b41-uuid')).toBe(false)
    expect(isSessionId('claude-code:')).toBe(false)
  })
})

describe('enc/dec session id segment', () => {
  const ids = [
    'claude-code:a1b2c3d4-5e6f-7890-abcd-ef1234567890',
    // legacy Claude path-derived key — contains `/`
    'claude-code:-home-rivet-rivetOS/a1b2c3d4-5e6f',
    // native id containing `:`
    'grok-build:sess_01HZX:sub:3',
    // both, plus non-ASCII
    'kimi-code:proj/名前:7f3a',
    'hermes:9b41-uuid',
  ]

  it('round-trips ids containing ":" and "/"', () => {
    for (const id of ids) {
      const seg = encodeSessionIdSegment(id)
      expect(decodeSessionIdSegment(seg)).toBe(id)
    }
  })

  it('emits unpadded base64url — safe as a single path segment', () => {
    for (const id of ids) {
      const seg = encodeSessionIdSegment(id)
      expect(seg).toMatch(/^[A-Za-z0-9_-]+$/)
      expect(seg).not.toContain('=')
      expect(seg).not.toContain('/')
      expect(seg).not.toContain('+')
    }
  })

  it('rejects encoding a non-canonical id', () => {
    expectInvalid(() => encodeSessionIdSegment('claude:abc'))
    expectInvalid(() => encodeSessionIdSegment(' claude-code:abc'))
  })

  it('rejects segments that are padded, non-base64url, or decode to garbage', () => {
    expectInvalid(() => decodeSessionIdSegment(''))
    expectInvalid(() => decodeSessionIdSegment('not base64url!'))
    // padded standard base64 of an otherwise valid id
    const padded = btoa('claude-code:abcd')
    expect(padded).toContain('=')
    expectInvalid(() => decodeSessionIdSegment(padded))
    // well-formed base64url that decodes to a non-canonical id
    expectInvalid(() => decodeSessionIdSegment(encodeToSegment('claude:abc')))
    expectInvalid(() => decodeSessionIdSegment(encodeToSegment('bare-uuid')))
  })

  it('rejects a base64url-shaped segment of impossible length (atob throws)', () => {
    // Passes the charset regex but length % 4 === 1 is not decodable — atob
    // raises a DOMException, which must surface as the typed code, not escape.
    expectInvalid(() => decodeSessionIdSegment('A'))
    expectInvalid(() => decodeSessionIdSegment('Y2xhdWRlLWNvZGU6YWJjZQ'.slice(0, 17)))
  })

  it('rejects a segment whose bytes are not valid UTF-8', () => {
    // 'claude-code:' + a lone 0xFF — decodable base64url, undecodable UTF-8.
    const segment = btoa(`claude-code:${String.fromCharCode(0xff)}`)
      .replace(/\+/g, '-')
      .replace(/\//g, '_')
      .replace(/=+$/, '')
    expect(segment).toMatch(/^[A-Za-z0-9_-]+$/)
    expectInvalid(() => decodeSessionIdSegment(segment))
  })
})

/** Encode without the canonical-id validation, to build bad-input fixtures. */
function encodeToSegment(raw: string): string {
  const bytes = new TextEncoder().encode(raw)
  let binary = ''
  for (const byte of bytes) binary += String.fromCharCode(byte)
  return btoa(binary).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '')
}

File: apps/rivet-android/app/src/test/java/dev/rivet/app/data/harness/HarnessSessionIdTest.kt
package dev.rivet.app.data.harness

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Test

/**
 * SessionId codec — parse on the FIRST colon, and `enc()`/`dec()` round-trip
 * for ids containing `:` and `/`, which is exactly why the segment is
 * base64url and not percent-encoding.
 *
 * Mirrors `packages/types/src/harness-session-id.test.ts`.
 */
class HarnessSessionIdTest {

    private fun expectInvalid(block: () -> Unit) {
        try {
            block()
            fail("expected InvalidSessionIdException")
        } catch (_: InvalidSessionIdException) {
            // expected
        }
    }

    // ---- parse ---------------------------------------------------------------

    @Test
    fun `parses a canonical id`() {
        val parsed = HarnessSessionIds.parse("claude-code:a1b2c3d4")
        assertEquals("claude-code", parsed.harnessId)
        assertEquals("a1b2c3d4", parsed.nativeSessionId)
    }

    @Test
    fun `splits on the first colon only`() {
        val parsed = HarnessSessionIds.parse("grok-build:sess:01HZX:tail")
        assertEquals("grok-build", parsed.harnessId)
        assertEquals("sess:01HZX:tail", parsed.nativeSessionId)
    }

    @Test
    fun `keeps slashes in the native half`() {
        val parsed = HarnessSessionIds.parse("claude-code:home-rivet-proj/9f2c-uuid")
        assertEquals("home-rivet-proj/9f2c-uuid", parsed.nativeSessionId)
    }

    @Test
    fun `rejects unknown harness ids, whitespace and empty halves`() {
        expectInvalid { HarnessSessionIds.parse("claude:abc") }
        expectInvalid { HarnessSessionIds.parse("cc:abc") }
        expectInvalid { HarnessSessionIds.parse(" claude-code:abc") }
        expectInvalid { HarnessSessionIds.parse("claude-code:abc ") }
        expectInvalid { HarnessSessionIds.parse("claude-code:") }
        expectInvalid { HarnessSessionIds.parse(":abc") }
        expectInvalid { HarnessSessionIds.parse("bare-uuid") }
    }

    @Test
    fun `parseOrNull and isSessionId agree with parse`() {
        assertNull(HarnessSessionIds.parseOrNull("bare-uuid"))
        assertFalse(HarnessSessionIds.isSessionId("bare-uuid"))
        assertTrue(HarnessSessionIds.isSessionId("hermes:9b41-uuid"))
        assertEquals("9b41-uuid", HarnessSessionIds.nativeIdOf("hermes:9b41-uuid"))
        assertNull(HarnessSessionIds.nativeIdOf("9b41-uuid"))
    }

    @Test
    fun `format validates and round-trips through parse`() {
        val id = HarnessSessionIds.format("kimi-code", "c7f2-uuid")
        assertEquals("kimi-code:c7f2-uuid", id)
        assertEquals("c7f2-uuid", HarnessSessionIds.parse(id).nativeSessionId)
        expectInvalid { HarnessSessionIds.format("claude", "abc") }
        expectInvalid { HarnessSessionIds.format("claude-code", "") }
    }

    // ---- enc / dec -----------------------------------------------------------

    @Test
    fun `enc dec round-trips ids containing a colon`() {
        val id = "grok-build:sess:01HZXABCDEF"
        val segment = HarnessSessionIds.encodeSegment(id)
        assertTrue(segment.matches(Regex("^[A-Za-z0-9_-]+$")))
        assertEquals(id, HarnessSessionIds.decodeSegment(segment))
    }

    @Test
    fun `enc dec round-trips ids containing a slash`() {
        val id = "claude-code:-home-rivet-projects/9f2c-4b1a-uuid"
        val segment = HarnessSessionIds.encodeSegment(id)
        assertFalse(segment.contains("/"))
        assertFalse(segment.contains("+"))
        assertFalse(segment.contains("="))
        assertEquals(id, HarnessSessionIds.decodeSegment(segment))
    }

    @Test
    fun `enc dec round-trips non-ascii native ids`() {
        val id = "hermes:会話-☃-01"
        assertEquals(id, HarnessSessionIds.decodeSegment(HarnessSessionIds.encodeSegment(id)))
    }

    @Test
    fun `enc is unpadded base64url of the utf-8 bytes`() {
        // 'claude-code:abcd' is 16 bytes, so standard base64 would pad.
        val segment = HarnessSessionIds.encodeSegment("claude-code:abcd")
        assertEquals("Y2xhdWRlLWNvZGU6YWJjZA", segment)
    }

    @Test
    fun `enc refuses a non-canonical id`() {
        expectInvalid { HarnessSessionIds.encodeSegment("claude:abc") }
        expectInvalid { HarnessSessionIds.encodeSegment("bare-uuid") }
        expectInvalid { HarnessSessionIds.encodeSegment(" claude-code:abc") }
    }

    @Test
    fun `dec refuses padded, non-base64url and non-canonical payloads`() {
        expectInvalid { HarnessSessionIds.decodeSegment("") }
        expectInvalid { HarnessSessionIds.decodeSegment("not base64url!") }
        // A padded encoding of a perfectly good id is still rejected.
        val padded = java.util.Base64.getUrlEncoder()
            .encodeToString("claude-code:abcd".toByteArray(Charsets.UTF_8))
        assertTrue(padded.endsWith("="))
        expectInvalid { HarnessSessionIds.decodeSegment(padded) }
        // Decodes cleanly, but not to a canonical SessionId.
        val bare = java.util.Base64.getUrlEncoder().withoutPadding()
            .encodeToString("bare-uuid".toByteArray(Charsets.UTF_8))
        expectInvalid { HarnessSessionIds.decodeSegment(bare) }
    }

    @Test
    fun `dec refuses bytes that are not valid utf-8`() {
        val segment = java.util.Base64.getUrlEncoder().withoutPadding()
            .encodeToString(byteArrayOf(0xC3.toByte(), 0x28))
        expectInvalid { HarnessSessionIds.decodeSegment(segment) }
    }

    // ---- segment selection ---------------------------------------------------

    @Test
    fun `segmentFor encodes canonical ids and passes bare native ids through`() {
        val canonical = "claude-code:9f2c-uuid"
        assertEquals(
            HarnessSessionIds.encodeSegment(canonical),
            HarnessSessionIds.segmentFor(canonical),
        )
        // The documented legacy shape: no harness prefix to encode, so it rides
        // as a plain percent-encoded segment (gateway-client parity).
        assertEquals("9f2c-uuid", HarnessSessionIds.segmentFor("9f2c-uuid"))
    }

    @Test
    fun `urlEncode escapes spaces as percent-20 not plus`() {
        assertEquals("a%20b", HarnessSessionIds.urlEncode("a b"))
        assertEquals("a%2Fb", HarnessSessionIds.urlEncode("a/b"))
    }

    @Test
    fun `roster commands are labels, never key material`() {
        assertEquals("claude", HarnessIds.rosterCommand(HarnessIds.CLAUDE_CODE))
        assertEquals("grok", HarnessIds.rosterCommand(HarnessIds.GROK_BUILD))
        assertEquals("kimi", HarnessIds.rosterCommand(HarnessIds.KIMI_CODE))
        assertEquals("hermes", HarnessIds.rosterCommand(HarnessIds.HERMES))
        assertEquals(4, HarnessIds.ALL.size)
    }
}
